src/main/java/com/squareup/kotlinpoet/TypeSpec.kt

```diff
@@ -441,6 +444,14 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
     fun addSuperinterface(superinterface: KClass<*>)
         = addSuperinterface(superinterface.asTypeName())
 
+    fun addSuperinterface(superinterface: ParameterSpec) = apply {
```

Comment:
> you'll still have to pass the actual ParameterSpec with the primary constructor definition.

Ok, now I see where you're coming from.  I don't think I explained my thinking behind the way I wrote it.  

When I thought of `superinterfaceBy(param: ParamSpec)`,  I thought it carried the implication that it'd be generated as a param of the primary constructor - even though the function call is on the `TypeSpec` with no accompanying `FunSpec`.  Since there's no distinction between a primary constructor and functions (at the moment), my reasoning was that the only way to support a configurable property (which is also delegated to by a type) was to leave it up to the caller to know that there needed to be a primary ctor for that type - and the delegate parameter would be added whether explicitly added or not. That's why I checked for a non-null constructor on the call to `build()`, and then the ParamSpec is added to the ctor if not done already. 

I'm fine with changing it though. My use case (which led to the pull request) has a default value for the parameter, which is why I went at the problem from that angle.  Maybe there needs to be a more specific type like `ConstructorSpec` in order to make that intuitive, but it's not my call to change up the type hierarchy